bert/src/deserializers.rs

```diff
@@ -0,0 +1,325 @@
+    type Error = Error;
+
+    forward_to_deserialize! {
+        bool usize u8 u16 u32 u64 isize i8 i16 i32 i64 f32 f64 char
+        str string unit seq seq_fixed_size bytes option map unit_struct
+        tuple_struct struct struct_field tuple ignored_any newtype_struct
+    }
+
+    #[inline]
+    fn deserialize<V: Visitor>(&mut self, visitor: V) -> Result<V::Value> {
+        if self.header.is_none() {
+            self.header = Some(try!(self.read_u8()));
+        }
+
+        let result = self.parse_value(visitor);
+        self.header = None;
+        result
+    }
+
+    #[inline]
+    fn deserialize_enum<V: EnumVisitor>(
+        &mut self, _enum: &'static str, _variants: &'static [&'static str],
+        mut _visitor: V
+    ) -> Result<V::Value> {
+        Err(Error::UnsupportedType)
+    }
+}
+
+
+struct BinarySeqVisitor<'a, R: 'a + Read> {
+    de: &'a mut Deserializer<R>,
+    length: Option<usize>
+}
+
+
+impl<'a, R: 'a + Read> BinarySeqVisitor<'a, R> {
+    #[inline]
+    fn new(de: &'a mut Deserializer<R>, length: Option<usize>) -> Self {
+        BinarySeqVisitor { de: de, length: length }
+    }
+}
+
+
+impl<'a, R: Read> de::Deserializer for BinarySeqVisitor<'a, R> {
+    type Error = Error;
+
+    fn deserialize<V>(&mut self, mut visitor: V) -> Result<V::Value>
+        where V: Visitor
+    {
+        visitor.visit_u8(try!(self.de.read_u8()))
+    }
+
+    forward_to_deserialize! {
+        bool usize u8 u16 u32 u64 isize i8 i16 i32 i64 f32 f64 char str string
+        unit option seq seq_fixed_size bytes map unit_struct newtype_struct
+        tuple_struct struct struct_field tuple enum ignored_any
+    }
+}
+
+
+impl<'a, R: Read> de::SeqVisitor for BinarySeqVisitor<'a, R> {
+    type Error = Error;
+
+    fn visit<T: Deserialize>(&mut self) -> Result<Option<T>> {
+        match self.length {
+            Some(0) => return Ok(None),
+            Some(ref mut len) => *len -= 1,
+            None => {}
+        };
+        Deserialize::deserialize(self).map(Some)
+    }
+
+    fn end(&mut self) -> Result<()> {
+        if let Some(0) = self.length {
+            Ok(())
+        } else {
+            Err(Error::TrailingBytes)
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        match self.length {
+            Some(len) => (len, self.length),
+            None => (0, Some(0))
+        }
+    }
+}
+
+
+struct BertBigNumberVisitor<'a, R: 'a + Read> {
+    de: &'a mut Deserializer<R>,
+    header: u8,
+}
+
+
+impl<'a, R: 'a + Read> BertBigNumberVisitor<'a, R> {
+    #[inline]
+    fn new(de: &'a mut Deserializer<R>, header: u8) -> Self {
+        BertBigNumberVisitor { de: de, header: header }
+    }
+}
+
+
+impl<'a, R: 'a + Read> de::Visitor for BertBigNumberVisitor<'a, R> {
+    type Value = BertBigInteger;
+
+    #[inline]
+    fn visit_newtype_struct<D>(&mut self, _: &mut D) -> StdResult<BertBigInteger, D::Error>
+        where D: de::Deserializer
+    {
+        let n = match self.header {
```

Comment:
You should not be accessing the deserializer directly. Instead, use `Deserialize::deserialize` calls to obtain the correct value, in this case, `let v: Vec<u8> = Deserialize::deserialize(deserializer)`